src/Analyzers/MSTest.Analyzers/TypeContainingTestMethodShouldBeATestClassAnalyzer.cs

```diff
@@ -0,0 +1,104 @@
+    private static readonly LocalizableResourceString MessageFormat = new(nameof(Resources.TypeContaningTestMethodShouldBeATestClassMessageFormat), Resources.ResourceManager, typeof(Resources));
+
+    internal static readonly DiagnosticDescriptor TypeContainingTestMethodShouldBeATestClassRule = DiagnosticDescriptorHelper.Create(
+        DiagnosticIds.TypeContainingTestMethodShouldBeATestClassRuleId,
+        Title,
+        MessageFormat,
+        Description,
+        Category.Usage,
+        DiagnosticSeverity.Info,
+        isEnabledByDefault: true);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+        = ImmutableArray.Create(TypeContainingTestMethodShouldBeATestClassRule);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(context =>
+        {
+            if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingTestMethodAttribute, out INamedTypeSymbol? testMethodAttributeSymbol)
+                && context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftVisualStudioTestToolsUnitTestingTestClassAttribute, out INamedTypeSymbol? testClassAttributeSymbol))
+            {
+                INamedTypeSymbol? taskSymbol = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask);
+                INamedTypeSymbol? valueTaskSymbol = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksValueTask);
+                bool canDiscoverInternals = context.Compilation.CanDiscoverInternals();
+                context.RegisterSymbolAction(
+                    context => AnalyzeSymbol(context, testClassAttributeSymbol, testMethodAttributeSymbol),
+                    SymbolKind.NamedType);
+            }
+        });
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol testClassAttributeSymbol, INamedTypeSymbol testMethodAttributeSymbol)
+    {
+        var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+        if (namedTypeSymbol.TypeKind != TypeKind.Class
+            || namedTypeSymbol.IsAbstract)
+        {
+            return;
+        }
+
+        bool isTestClass = false;
+        foreach (AttributeData classAttribute in namedTypeSymbol.GetAttributes())
+        {
+            if (classAttribute.AttributeClass.Inherits(testClassAttributeSymbol))
+            {
+                isTestClass = true;
+                break;
+            }
+        }
+
+        if (isTestClass)
+        {
+            return;
+        }
+
+        bool hasTestMethod = false;
+        INamedTypeSymbol? currentType = namedTypeSymbol;
+        do
+        {
+            foreach (ISymbol classMember in currentType.GetMembers())
```

Comment:
Only methods can be marked with `[TestMethod]` so you should filter out to ensure you don't call `GetAttributes` on all members.